#### Problem 9

A Pythagorean triplet is a set of three natural numbers, $a < b < c$, for which,

$a^2 + b^2 = c^2$

For example, $3^2 + 4^2 = 9 + 16 = 25 = 5^2$.

There exists exactly one Pythagorean triplet for which $a + b + c = 1000$.
Find the product $abc$.

In [1]:
# Check if numbers form a pythagorean triplet.
def is_pythagorean_triplet(a, b, c):
    return (a**2 + b**2 == c**2)

In [2]:
# Search through all possible combinations.
# 0 < a < target_sum / 3 since a < b < c
# a < b < target_sum / 2 since b < c
# c = target_sum - a - b since a + b + c = target_sum
def search_numbers(target_sum):
    
    # Store all solutions in list
    solutions = []
    
    # 0 < a < target_sum / 3 since a < b < c
    for a in range(1, int((target_sum)/3)+1):
        
        # a < b < target_sum / 2 since b < c
        for b in range(a+1, int((target_sum+1)/2)+1):
            
            # c = target_sum - a - b since a + b + c = target_sum
            c = target_sum - a - b
            
            # Check if the set is a pythagorean triple.
            if is_pythagorean_triplet(a, b, c): solutions.append([a, b, c])
    
    # Return all solutions.
    return solutions

In [3]:
target_sum = 1000
problem_solution = search_numbers(target_sum)[0]

In [4]:
print(f'The Pythagorean triple that sums to {target_sum:,}:')
print(',    '.join(str(number) for number in problem_solution))
print(f'{problem_solution[0]}   + {problem_solution[1]}   + {problem_solution[2]}   = {problem_solution[0] + problem_solution[1] + problem_solution[2]:,}')
print(f'{problem_solution[0]}^2 + {problem_solution[1]}^2 + {problem_solution[2]}^2 = {problem_solution[0]**2 + problem_solution[1]**2 + problem_solution[2]**2:,}')
print(f'{problem_solution[0]}   * {problem_solution[1]}   * {problem_solution[2]}   = {problem_solution[0] * problem_solution[1] * problem_solution[2]:,}')

The Pythagorean triple that sums to 1,000:
200,    375,    425
200   + 375   + 425   = 1,000
200^2 + 375^2 + 425^2 = 361,250
200   * 375   * 425   = 31,875,000


Problem 9 Solution:

The Pythagorean triple that sums to 1,000:

$200,    375,    425$

$200   + 375   + 425   = 1,000$

$200^2 + 375^2 + 425^2 = 361,250$

$200   * 375   * 425   = 31,875,000$.